python/runtime/cudaq/qis/py_qubit_qis.cpp

```diff
@@ -0,0 +1,344 @@
+/*************************************************************** -*- C++ -*- ***
+ * Copyright (c) 2022 - 2023 NVIDIA Corporation & Affiliates.                  *
+ * All rights reserved.                                                        *
+ *                                                                             *
+ * This source code and the accompanying materials are made available under    *
+ * the terms of the Apache License 2.0 which accompanies this distribution.    *
+ *******************************************************************************/
+
+#include "py_qubit_qis.h"
+#include "cudaq/qis/qubit_qis.h"
+#include <fmt/core.h>
+#include <pybind11/stl.h>
+
+namespace cudaq {
+
+namespace details {
+
+/// @brief Analyze the incoming arguments and map them to a vector
+/// of qubit indices.
+std::vector<std::size_t> analyzeAndHandlePyArgs(py::args &args) {
+  if (args.size() == 1) {
+    // Is this arg a qubit or a qvector/qview?
+    if (py::isinstance<qubit &>(args[0])) {
+      return {args[0].cast<qubit &>().id()};
+    }
+
+    // This has to be a broadcast over a register
+    std::vector<std::size_t> mapped;
+    if (py::isinstance<qvector<> &>(args[0])) {
+      auto &casted = args[0].cast<qvector<> &>();
+      std::transform(casted.begin(), casted.end(), std::back_inserter(mapped),
+                     [](auto &&el) { return el.id(); });
+    } else if (py::isinstance<qview<> &>(args[0])) {
+      auto &casted = args[0].cast<qview<> &>();
+      std::transform(casted.begin(), casted.end(), std::back_inserter(mapped),
+                     [](auto &&el) { return el.id(); });
+    }
+
+    return mapped;
+  }
+
+  // There are multiple args here
+  std::vector<std::size_t> mapped;
+  for (auto &arg : args) {
+    if (py::isinstance<qubit &>(arg))
+      mapped.push_back(arg.cast<qubit &>().id());
+    else if (py::isinstance<qvector<> &>(arg)) {
+      auto &casted = arg.cast<qvector<> &>();
+      std::transform(casted.begin(), casted.end(), std::back_inserter(mapped),
+                     [](auto &&el) { return el.id(); });
+    } else if (py::isinstance<qview<> &>(arg)) {
+      auto &casted = arg.cast<qview<> &>();
+      std::transform(casted.begin(), casted.end(), std::back_inserter(mapped),
+                     [](auto &&el) { return el.id(); });
+    } else
+      throw std::runtime_error("Invalid type passed to a quantum operation.");
+  }
+
+  return mapped;
+}
```

Comment:
Would it be possible to merge the two code paths here, i.e., just keep the `for (auto &arg : args)` path?

Reply:
Yep sure can, good catch.